crates/block-producer/src/errors.rs

```diff
@@ -100,28 +101,59 @@ pub enum AddTransactionError {
-                VerifyTxError::InputNotesAlreadyConsumed(_)
-                | VerifyTxError::UnauthenticatedNotesNotFound(_)
-                | VerifyTxError::OutputNotesAlreadyExist(_)
-                | VerifyTxError::IncorrectAccountInitialCommitment { .. }
-                | VerifyTxError::InvalidTransactionProof(_),
-            )
-            | AddTransactionError::Expired { .. }
-            | AddTransactionError::TransactionDeserializationFailed(_) => {
-                Self::invalid_argument(value.as_report())
-            },
+// Error codes for gRPC Status::details
+// =================================================================================================
 
-            // Internal errors which should not be communicated to the user.
-            AddTransactionError::VerificationFailed(VerifyTxError::StoreConnectionFailed(_))
-            | AddTransactionError::StaleInputs { .. } => Self::internal("Internal error"),
+impl AddTransactionError {
+    fn api_error(&self) -> SubmitProvenTransactionError {
+        match self {
+            AddTransactionError::VerificationFailed(tx_verify_error) => match tx_verify_error {
+                VerifyTxError::InputNotesAlreadyConsumed(_) => {
+                    SubmitProvenTransactionError::InputNotesAlreadyConsumed
+                },
+                VerifyTxError::StoreConnectionFailed(_) => {
+                    SubmitProvenTransactionError::InternalError
+                },
+                VerifyTxError::UnauthenticatedNotesNotFound(_) => {
+                    SubmitProvenTransactionError::UnauthenticatedNotesNotFound
+                },
+                VerifyTxError::OutputNotesAlreadyExist(_) => {
+                    SubmitProvenTransactionError::OutputNotesAlreadyExist
+                },
+                VerifyTxError::IncorrectAccountInitialCommitment { .. } => {
+                    SubmitProvenTransactionError::IncorrectAccountInitialCommitment
+                },
+                VerifyTxError::InvalidTransactionProof(_) => {
+                    SubmitProvenTransactionError::InvalidTransactionProof
+                },
+            },
+            AddTransactionError::StaleInputs { .. } => SubmitProvenTransactionError::InternalError,
+            AddTransactionError::Expired { .. } => SubmitProvenTransactionError::TransactionExpired,
+            AddTransactionError::TransactionDeserializationFailed(_) => {
+                SubmitProvenTransactionError::DeserializationFailed
+            },
         }
     }
 }
 
+impl From<AddTransactionError> for tonic::Status {
+    fn from(value: AddTransactionError) -> Self {
+        let api_error = value.api_error();
+
+        let message = if api_error.is_internal() {
+            "Internal error".to_owned()
+        } else {
+            value.as_report()
+        };
+
+        tonic::Status::with_details(
+            api_error.tonic_code(),
+            message,
+            vec![api_error.api_code()].into(),
```

Comment:
Is there an there an implicit assumption here that we know how protobuf enums are serialized? Or do we not care about this and just using protobuf messages as means of specifying numeric error values, but otherwise don't care about their serialization?

Reply:
@SantiagoPittella could you try figure out how we could instead write something like:

```rust
tonic::Status::with_details(
    api_error.tonic_code(),
    message,
    tonic::to_bytes(api_error)
)
```

Reply:
I found something but it is more like a workaround than a solution, i don't really like it. It is basically wrapping the enum in a message where it is the only field, which does implement serialization

Reply:
Not really sure how to tackle the serialization yet. Should we merge this PR with a comment on the serialization?